m2mresource: save RAM and ROM at M2MExecuteParameter

TeroJaasko opened the pull request:
Change the API, use and internal storage of M2MExecuteParameter.
No need to allocate temp object from heap and copy the value just
for the duration of execute callback. This also reduces code
footprint by 120B on GCC.

Note: this changes the user facing API.

TeroJaasko commented:
Just a simple drive-by fix. We need to evaluate, if this justifies a API break. If not, lets put the code behind MEMORY_OPTIMIZED_API and leave older one there with m2m_deprecated tags.

@yogpan01 @anttiylitokola @kuggenhoffen

yogpan01 commented:
@TeroJaasko Can you modify what we discussed ?

TeroJaasko commented:
Moved the new version behind the MEMORY_OPTIMIZED_API flag. Now only change to API is M2MExecuteParameter::get_argument_value() is "const uint8_t*" instead of "uint8_t*". But client should not try to modify the given resource value anyway.

yogpan01 commented:
@TeroJaasko Can you please check the unit tests ? There seems to be some compilation errors in that PR check.

JanneKiiskila commented:
Do we need to version the API, if the new optimized is not compatible with the older optimized?

TeroJaasko commented:
Fixed the unit tests, changed target to memory_optimizations branch and rebased against it. 
@JanneKiiskila; yes, technically we should have a MEMORY_OPTIMIZED_API_V2 or something, but is it really worth the trouble?

TeroJaasko commented:
To be rebased with master after #427 is merged.

TeroJaasko commented:
This needs a "git rebase master" after #498 is merged.
